ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Node/ProtocolVersions/Internal.hs

```diff
@@ -0,0 +1,187 @@
+cardanoMainnetProtocolVersions experimentalHardForksEnabled =
+    CardanoProtocolVersions {
+        protVerByron   = Byron.Update.ProtocolVersion 3 0 0
+      , protVerShelley = SL.ProtVer (SL.natVersion @3) 0
+      , protVerAllegra = SL.ProtVer (SL.natVersion @4) 0
+      , protVerMary    = SL.ProtVer (SL.natVersion @5) 0
+      , protVerAlonzo  = SL.ProtVer (SL.natVersion @7) 0
+      , protVerBabbage = if experimentalHardForksEnabled then
+                           SL.ProtVer (SL.natVersion @9) 0
+                         else
+                           SL.ProtVer (SL.natVersion @8) 0
+      , protVerConway  = SL.ProtVer (SL.natVersion @9) 0
+      }
+
+cardanoMaxProtocolVersion :: CardanoProtocolVersions c -> SL.ProtVer
+cardanoMaxProtocolVersion = maxProtocolVersion . getCardanoProtocolVersions
+
+{-------------------------------------------------------------------------------
+  HasProtocolVersion
+-------------------------------------------------------------------------------}
+
+-- | This type family is not an associated type of the 'HasProtocolVersion'
+-- class, as it reduces duplication: all Shelley blocks have the same
+-- 'ProtocolVersion' type.
+type ProtocolVersion :: Type -> Type
+type family ProtocolVersion blk where
+  ProtocolVersion ByronBlock               = Byron.Update.ProtocolVersion
+  ProtocolVersion (ShelleyBlock proto era) = SL.ProtVer
+  ProtocolVersion (CardanoBlock c)         = CardanoProtocolVersions c
+
+type HasProtocolVersion :: Type -> Constraint
+class HasProtocolVersion blk where
+  extractProtocolVersion :: ProtocolParams blk -> ProtocolVersion blk
+
+{-------------------------------------------------------------------------------
+  HasProtocolVersion: block instances
+-------------------------------------------------------------------------------}
+
+instance HasProtocolVersion ByronBlock where
+  extractProtocolVersion = byronProtocolVersion
+
+instance HasProtocolVersion (ShelleyBlock (TPraos c) (ShelleyEra c)) where
+  extractProtocolVersion = shelleyProtVer
+
+instance HasProtocolVersion (ShelleyBlock (TPraos c) (AllegraEra c)) where
+  extractProtocolVersion = allegraProtVer
+
+instance HasProtocolVersion (ShelleyBlock (TPraos c) (MaryEra c)) where
+  extractProtocolVersion = maryProtVer
+
+instance HasProtocolVersion (ShelleyBlock (TPraos c) (AlonzoEra c)) where
+  extractProtocolVersion = alonzoProtVer
+
+instance HasProtocolVersion (ShelleyBlock (Praos c) (BabbageEra c)) where
+  extractProtocolVersion = babbageProtVer
+
+instance HasProtocolVersion (ShelleyBlock (Praos c) (ConwayEra c)) where
+  extractProtocolVersion = conwayProtVer
+
+{-------------------------------------------------------------------------------
+  Value for /each/ era
+-------------------------------------------------------------------------------}
+
+type PerEraProtocolVersion :: [Type] -> Type
+newtype PerEraProtocolVersion xs = PerEraProtocolVersion {
```

Comment:
Is this intermediate type buying us something here, or is it just there for consistency with other "per-era" `NP`s?